core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1668,12 +1668,19 @@ public static boolean isAtomic(RelDataType type) {
         || SqlTypeUtil.isBoolean(type);
   }
 
-  /** Returns a DECIMAL type with the maximum precision/scale for the current
+  /** Returns a DECIMAL type with the maximum precision for the current
    * type system. */
   public static RelDataType getMaxPrecisionScaleDecimal(RelDataTypeFactory factory) {
     int maxPrecision = factory.getTypeSystem().getMaxNumericPrecision();
     int maxScale = factory.getTypeSystem().getMaxNumericScale();
 
+    // maxScale should not greater than maxPrecision.
+    // If they are equal, e.g. Decimal(19,19)
+    // means we can only have decimal places.
+    while (maxScale >= maxPrecision) {
+      maxScale = maxScale / 2;
+    }
```

Comment:
I'm afraid not all the sql engines have the rule that `maxScale = maxScale / 2`, we should add a check for the validity instead of modifying the value silently. And i'm inclined to fix the default max precision/scale for Calcite type system.

Reply:
> I'm afraid not all the sql engines have the rule that `maxScale = maxScale / 2`, we should add a check for the validity instead of modifying the value silently. And i'm inclined to fix the default max precision/scale for Calcite type system.

This is also a proposal, I'll take a look at Hive/Spark decimal's max scale/repcision, but I'm afraid they may be the same, or we don't get max but get default?

Reply:
### Spark
**Default**: (10, 0)
**Max**: The precision can be up to 38, scale can also be up to 38 (less or equal to precision).

Reply:
I got the impression that many systems support max precision scale as `decimal(38, 18)`.

Reply:
The "getMaxPrecisionScaleDecimal" means suppose to get a max Decimal, but the Decimal(max, max) is not this meaning.

Reply:
> I got the impression that many systems support max precision scale as `decimal(38, 18)`.

Hi @danny0405, IMO the specific number here is not so important (like18  or 19), we just need return a big decimal here, so can we finalize a number just like Decimal(38, 18)?

Reply:
cc @julianhyde

Reply:
Let's just make the valid max scale half of the max precision, e.g. the` max scale = max precision/2`.

Reply:
> Let's just make the valid max scale half of the max precision, e.g. the` max scale = max precision/2`.

OK